### ISSUE TITLE
Bundles absolute path from dev machine into built artifacts

### ISSUE THREAD

**Issue by joshribakoff** [prisma/prisma#18108]
### Bug description

Trying to bundle an application and deploy to AWS lambda. When I generate the Prisma client, it create a file `index.js` that contains:

```
"output": {
      "value": "/Users/joshribakoff/lift-tracker/prisma/generated/client",
      "fromEnvVar": null
    },
```

Obviously this path is not going to exist on the AWS Lambda machine 🤦 :

```
2023-02-26 14:21:31.137 ERROR   Unhandled Promise Rejection     {"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: Query engine library for current platform \"rhel-openssl-1.0.x\" could not be found.\nYou incorrectly pinned it to rhel-openssl-1.0.x\n\nThis probably happens, because you built Prisma Client on a different platform.\n(Prisma Client looked in \"/var/task/dist/packages/api/libquery_engine-rhel-openssl-1.0.x.so.node\")\n\nSearched Locations:\n\n  /var/task/.prisma/client\n  /Users/joshribakoff/lift-tracker/prisma/generated/client\n  /var/task/dist/packages\n  /var/task/dist/packages/api\n  /var/task/dist\n  /tmp/prisma-engines\n 
```

The workaround is to copy it to one of the relative paths it's checking:

```
path.join(__dirname, "libquery_engine-darwin.dylib.node");
path.join(process.cwd(), "prisma/generated/client/libquery_engine-darwin.dylib.node")

path.join(__dirname, "libquery_engine-rhel-openssl-1.0.x.so.node");
path.join(process.cwd(), "prisma/generated/client/libquery_engine-rhel-openssl-1.0.x.so.node")
path.join(__dirname, "schema.prisma");
path.join(process.cwd(), "prisma/generated/client/schema.prisma")

```

Then it finds it. But the fact Prisma is taking information about my (private) local development environment and bundling that into production code is not only unneccassary, it's a potential information leak or at the very least confusing and does not instill confidence that the Prisma team has a handle on things.

### How to reproduce

n/a

### Prisma information

n/a

### Environment & setup

Mac, Node 18

### Prisma Version

```
"@prisma/client": "3.15.2",
"prisma": "3.15",
```

**Comment by millsp** [comment]
Hey @joshribakoff sorry to hear this, we'd like to make things better in the future.
Would you be able to provide us with some information about you setup?
* Which bundler are you using?
* Which framework are you using?
* How do you deploy to AWS?

If you have the time, we would appreciate a reproduction that represents your project the best. These are areas of our code that we are currently wanting to improve, so any help from your side is very welcome.

**Comment by joshribakoff** [comment]
@millsp thanks!

nx + webpack + serverless framework

let me know if you can’t replicate, but I think that constructively the issue here is two fold

- Things will be much simpler if you simply checked in one place for an expected file. That one place should either be relative path or easily overridden with an environment, variable. It is not clear right now where Prisma is getting a lot of these paths from.
- It would also again be easier to debug if the error message did not insinuate that I don’t know what I’m doing. Simply tell me you didn’t find the file in the path that you looked for it in. Don’t presuppose I’ve pinned the wrong version because that sends me down the wrong path.


Thank you for considering my feedback! Let me know how i can help

**Comment by joshribakoff** [comment]
By the way in case it wasnt clear i got it working by configuring nx to add prisma binary in the “assets” in a manner where it gets copied into the “dist” folder that the serverless framework zips up and uploads to aws.

its just that the error message is confusing and its confusing im uploading code that references my mac’s filesystem. I would claim we could do better :)

**Comment by janpio** [comment]
We unfortunately can't just check one expected place, as the file can be in multiple places depending on the setup of the project and deployment. Equally, the error message used to cover the most common cases where this went wrong - which just does not apply to you. We have an issue open already to fix both the location logic, and the error message.

**Comment by joshribakoff** [comment]
> We unfortunately can't just check one expected place, as the file can be in multiple places depending on the setup of the project and deployment.

well the aws lambda is certainly not going to find it referencing my local mac :) couldnt it just run the checks to determine the location prior to building the generated code? As mentioned an env var would provide a simple way to override the value for deployment 

> We have an issue open already to fix both the location logic, and the error message.

awesome! Im still a bit confused though because above it seems like you implied the location checking logic is fine as is. Got a link to those issues?

**Comment by janpio** [comment]
>  Im still a bit confused though because above it seems like you implied the location checking logic is fine as is.

I just mentioned that it exists for a good reason. I did not want to imply that it is without faults or bugs, and especially the error messages we know are terrible. I would never use the word "fine" in context to that logic 😆 It usually does not cause issues though that can not be overcome (as you also figured out).

> Got a link to those issues?

I do not have the issues handy, but we will look into the whole cluster of https://github.com/prisma/prisma/issues?q=is%3Aopen+label%3A%22topic%3A+Query+engine+...+could+not+be+found.%22+sort%3Aupdated-desc soon and those will be/become one of those. Trust me when I say that multiple people want to really, fully clean that up.

**Comment by joshribakoff** [comment]
Makes sense! Thanks for taking the time to triage this and chat it over <3 i disagree it needs to check multiple locations and claim 1 location (that the user can override) should work, something like looking for the binary in the same folder as the generated client by concatenation of __dirname and a relative path (and allowing me to fully override that with my own absolute path via an env var).

then ideally the error just says it was not found in X location and links to the docs about generating the client and configuration of the binary

thanks again! I trust youll take the liberty of closing this if it helps you organizing duplicates

**Comment by beanaroo** [comment]
We noticed our builds were no longer deterministic after introducing prisma. This would cause the entire project to redeploy unnecessarily due to bundle hash differences. After a lot of troubleshooting, we discovered that this was due to `config.generator.output.value` being different for every build.

![image](https://github.com/prisma/prisma/assets/1748909/83ab9589-37ec-4662-81fa-016eacb8e276)

I'd like to ask if the prisma team can consider altering this output behaviour or potentially making it configurable as it is importance for adopting [reproducible builds](https://reproducible-builds.org/) and being able to orchestrate/optimise deployments.

To work around this, we have prepared an esbuild plugin:

```javascript
const rewritePrismaConfigOutputPath = {
  name: 'rewrite-prisma-config-output-path',
  setup(build) {
    build.onStart(() => {
      // rewrite the absolute path found in the .prisma/client index file
      const prismaClientIndexFile = 'node_modules/.prisma/client/index.js';
      fs.readFile(prismaClientIndexFile, 'utf8', function (err, data) {
        if (err) {
          return console.log(err);
        }
        const result = data.replace(/\/.*\/(node_modules\/\@prisma\/client)/g, '$1');

        fs.writeFile(prismaClientIndexFile, result, 'utf8', function (err) {
          if (err) return console.log(err);
        });
      });
    });
  },
};
```
